presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestFullOuterJoinWithCoalesce.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.sql.planner.assertions.BasePlanTest;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.aggregation;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.anyTree;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.equiJoinClause;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.exchange;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.expression;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.join;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.PARTIAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
+
+public class TestFullOuterJoinWithCoalesce
+        extends BasePlanTest
+{
+    @Test
+    public void testFullOuterJoinWithCoalesce()
+    {
+        assertDistributedPlan("SELECT coalesce(r.a, ts.a) " +
+                "FROM (" +
+                "   SELECT coalesce(t.a, s.a) AS a " +
+                "   FROM (VALUES (1), (2), (3)) t(a) " +
+                "   FULL OUTER JOIN (VALUES (1), (4)) s(a)" +
+                "   ON t.a = s.a) ts " +
+                "FULL OUTER JOIN (VALUES (2), (5)) r(a) " +
+                "ON ts.a = r.a",
+                anyTree(
+                        project(
+                                ImmutableMap.of("expr", expression("coalesce(r, ts)")),
```

Comment:
why `r` is on the left side, I mean why `coalesce(r, ts)` when predicate is `ON ts.a = r.a`?

Reply:
For full other join, if the input is partitioned on both `ts.a` and `r.a`, the output is partitioned on both `coalesce(ts.a, r.a)` and `coalesce(r.a, ts.a)`. The order of the coalesce should not matter.